src/storage/mutate/AddVerticesProcessor.cpp

```diff
@@ -108,132 +117,162 @@ void AddVerticesProcessor::process(const cpp2::AddVerticesRequest& req) {
+                        */
+                        if (nReader != nullptr) {
+                            auto ni = indexKey(partId, vid, nReader.get(), index);
+                            if (!ni.empty()) {
+                                // Check the index is building for the specified partition or not.
+                                auto indexState = env_->getIndexState(spaceId_, partId);
+                                if (env_->checkRebuilding(indexState)) {
+                                    auto opKey = OperationKeyUtils::modifyOperationKey(partId, ni);
+                                    batchHolder->put(std::move(opKey), "");
+                                } else if (env_->checkIndexLocked(indexState)) {
+                                    LOG(ERROR) << "The index has been locked: "
+                                               << index->get_index_name();
+                                    pushResultCode(cpp2::ErrorCode::E_CONSENSUS_ERROR, partId);
+                                    onFinished();
+                                    return;
+                                } else {
+                                    batchHolder->put(std::move(ni), "");
+                                }
+                            }
+                        }
+
+                        /*
+                        * step 3 , Insert new vertex data
+                        */
+                        batchHolder->put(std::move(key), std::move(retEnc.value()));
                     }
                 }
+                dummyLock.emplace_back(std::make_tuple(spaceId_, partId, tagId, vid));
+
+                if (FLAGS_enable_vertex_cache && vertexCache_ != nullptr) {
+                    vertexCache_->evict(std::make_pair(vid, tagId));
+                    VLOG(3) << "Evict cache for vId " << vid
+                            << ", tagId " << tagId;
+                }
+            }
+        }
+        auto batch = encodeBatchValue(std::move(batchHolder)->getBatch());
+        if (batch.empty()) {
+            handleAsync(spaceId_, partId, kvstore::ResultCode::SUCCEEDED);
+        } else {
+            nebula::MemoryLockGuard<VMLI> lg(env_->verticesML_.get(), dummyLock, true);
+            if (!lg) {
+                auto conflict = lg.conflictKey();
+                LOG(ERROR) << "vertex conflict "
+                           << std::get<0>(conflict) << ":"
+                           << std::get<1>(conflict) << ":"
+                           << std::get<2>(conflict) << ":"
+                           << std::get<3>(conflict);
+                pushResultCode(cpp2::ErrorCode::E_CONSENSUS_ERROR, partId);
+                onFinished();
+                return;
             }
+            auto callback = [partId, this](kvstore::ResultCode code) {
+                handleAsync(spaceId_, partId, code);
+            };
+            env_->kvstore_->asyncAppendBatch(spaceId_, partId, std::move(batch), callback);
```

Comment:
It seems `MemoryLockGuard` life cycle is short than `asyncAppendBatch`, would something go wrong?

Reply:
> It seems `MemoryLockGuard` life cycle is shorter than `asyncAppendBatch`, would something go wrong?
> I suppose unlock should happen in callback.
Good point!

Reply:
> It seems `MemoryLockGuard` life cycle is shorter than `asyncAppendBatch`, would something go wrong?
> I suppose unlock should happen in callback.

good point